File: mini-claw-code-starter-py/src/mini_claw_code_starter_py/tools/write.py
from __future__ import annotations

from typing import Any

from ..types import ToolDefinition


class WriteTool:
    def __init__(self) -> None:
        self._definition = None

    @property
    def definition(self) -> ToolDefinition:
        if self._definition is None:
            raise RuntimeError("WriteTool.new() must initialize self._definition")
        return self._definition

    @classmethod
    def new(cls) -> "WriteTool":
        raise NotImplementedError(
            "Define a write tool with required 'path' and 'content' string parameters"
        )

    async def call(self, args: Any) -> str:
        raise NotImplementedError(
            "Extract path and content, create parent directories, write the file, and confirm"
        )

File: mini-claw-code-starter-ts/src/tools/write.ts
import { mkdir, writeFile } from "node:fs/promises";

import type { JsonValue, Tool } from "../types";
import { ToolDefinition } from "../types";

export class WriteTool implements Tool {
  readonly toolDefinition: ToolDefinition;

  constructor() {
    this.toolDefinition = ToolDefinition.new(
      "write",
      "Write content to a file, creating directories as needed.",
    )
      .param("path", "string", "The file path to write to", true)
      .param("content", "string", "The content to write to the file", true);
  }

  static new(): WriteTool {
    return new WriteTool();
  }

  definition(): ToolDefinition {
    return this.toolDefinition;
  }

  async call(_args: JsonValue): Promise<string> {
    void mkdir;
    void writeFile;
    throw new Error(
      "TODO(ch4): extract path/content, create parent directories, write the file, and confirm success",
    );
  }
}
